presto-spi/src/main/java/com/facebook/presto/spi/plan/PlanNode.java

```diff
@@ -32,12 +32,23 @@
 {
     private final Optional<SourceLocation> sourceLocation;
     private final PlanNodeId id;
+    /**
+     * A plan node id, which is assigned by the Presto optimizer, to represent
+     * same canonical nodes throughout the planning phase.
+     * During planning, same canonical plan node id should be reused in a
+     * different plan node iff number of output rows/size remains similar.
+     * Once assigned by the optimizer, further optimizer rules should respect this id when changing the plan.
+     *
+     * For example, when doing pushdown: Filter(TableScan()) -> TableScan(),
+     * output TableScan should have same canonical id as input Filter node
+     */
+    private final Optional<PlanNodeId> canonicalId;
 
-    protected PlanNode(Optional<SourceLocation> sourceLocation, PlanNodeId id)
+    protected PlanNode(Optional<SourceLocation> sourceLocation, PlanNodeId id, Optional<PlanNodeId> canonicalId)
```

Comment:
maybe call it statisticsId or something like that since it's basically used for stats matching.